x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -603,6 +606,105 @@ protected LogicalPlan rule(OrderBy orderBy) {
         }
     }
 
+    /**
+     * Remove unused columns created in the plan, in fields inside eval or aggregations inside stats.
+     */
+    static class PruneColumns extends Rule<LogicalPlan, LogicalPlan> {
+
+        @Override
+        public LogicalPlan apply(LogicalPlan plan) {
+            var used = new Holder<>(new AttributeSet());
+            // don't remove Evals without any Project/Aggregate (which might not occur as the last node in the plan)
+            var seenProjection = new Holder<>(Boolean.FALSE);
+
+            // start top-to-bottom
+            // and track used references
+            var pl = plan.transformDown(p -> {
+                // skip nodes that simply pass the input through
+                if (p instanceof Limit) {
+                    return p;
+                }
+
+                // remember used
+                var usedSet = used.get();
+                boolean recheck;
+                // analyze the unused items against dedicated 'producer' nodes such as Eval and Aggregate
+                // perform a loop to retry checking if the current node is completely eliminated
+                do {
+                    recheck = false;
+                    if (p instanceof Aggregate aggregate) {
+                        var pruned = seenProjection.get() ? removeUnused(aggregate.aggregates(), usedSet) : null;
```

Comment:
optional: not sure if `pruned` means those removed or those kept after pruning. maybe `kept` might be more suggestive?